### DISCUSSION TITLE
4.38 unix socket and subparh

### DISCUSSION THREAD

**Discussion by cyberb** [authelia/authelia#6460] [Pre-Release Feedback]
Hi, I am very new to authelia and I am trying to set it up using nginx and unix socket address.
Unfortunately I do not have an option (right now) to have an auth.* domain and wanted to know if I could have authelia (portal and verify) under a directory.
Path property on 4.37 seems to serve exactly that purpose, why not have it on unix socket mode?

**Comment by cyberb** [comment]
Another reason for a directory mode is that I am already on an app web (transmission.mydevice.syncloud.it) which has a wildcard certificate (*.mydevice.syncloud.it) so adding next level domain breaks ssl (auth.transmission.mydevice.syncloud.it))

**Comment by james-d-elliott** [comment]
Because the unix socket file path uses the path portion of the parsed URL which makes up the address, they are two niche scenarios and I'm not convinced the effort or maintenance of those two features working together is either possible without reverting all the simplification we made in this area or going to be used by many people. 

People are still welcome to use the old convention for the time being (though that still wont work with the unix socket) or use the new one, subpath works fine with the address value provided it's not a unix socket.

**Comment by cyberb** [comment]
What about using something %2F approach, sometbing like this:
```
unix://%2Fpath%2Fto%2Ffile/subpath
```
I am not sure if url.Parse supporta it out of the box, but some simple custom parsing can be introduced here.

**Reply by james-d-elliott** [reply]
The intention behind the changes was to reduce our reliance on custom implementations of things. We use the underlying url.URL interface which holds compliant URI's and URL's, pretty sure that's not a valid URL as the first `%` must occur after the first `/` with the exclusion of the scheme. Why are you not interested in TCP?

**Comment by cyberb** [comment]
TCP, unfortunately in my case of a self-hosted app store (https://github.com/syncloud/platform) I have so many apps and each has many services which leads to tcp port management hell. Users install combinations of apps so I had to either use global list of unique ports to eliminate the conflict or rely on unix sockets to not have this problem. It may not be the best solution but a relatively easy one in my case.

**Reply by james-d-elliott** [reply]
The only option I can really think of then is a query or fragment parameter.. think we used that for another reason for the socket let me see.

**Comment by cyberb** [comment]
Also I think I have seen somethere people use this form, but may be that is for db conn string:
```
postgresql:///dbname?host=/var/lib/postgresql
postgresql://%2Fvar%2Flib%2Fpostgresql/dbname
```

**Comment by james-d-elliott** [comment]
Give https://github.com/authelia/authelia/pull/6463 a shot for me? Should be built in about 5 mins.

**Reply by james-d-elliott** [reply]
Ah.. one thing I've not accounted for, uhg. Give me a second.

**Reply by cyberb** [reply]
No rush, thanks for helping!

**Reply by james-d-elliott** [reply]
Give it a shot. Syntax for the unix socket `/var/run/authelia.sock` with the umask `0022` and traditional `path` of `/auth` should be: `unix:///var/run/authelia.sock?umask=0022&path=auth` (i.e. without the forward slash).

**Reply by cyberb** [reply]
Worked like a charm, thanks!

**Comment by cyberb** [comment]
Sorry tried actually to login and got this error, not sure it is related to this change or not:

```
time="2023-12-27T10:56:55Z" level=error msg="Failed to parse 1FA request body: unable to parse body: unexpected end of JSON input" method=POST path=/api/firstfactor remote_ip=192.168.128.3 stack="
github.com/authelia/authelia/v4/internal/handlers/handler_firstfactor.go:27 handleRouter.FirstFactorPOST.func9
github.com/authelia/authelia/v4/internal/middlewares/bridge.go:54           handleRouter.(*BridgeBuilder).Build.func5.1
github.com/authelia/authelia/v4/internal/middlewares/headers.go:56          SecurityHeadersCSPNone.func1
github.com/authelia/authelia/v4/internal/middlewares/headers.go:82          SecurityHeadersNoStore.func1
github.com/authelia/authelia/v4/internal/middlewares/headers.go:12          SecurityHeaders.func1
github.com/fasthttp/router@v1.4.22/router.go:420                            (*Router).Handler
github.com/authelia/authelia/v4/internal/middlewares/log_request.go:14      handleRouter.LogRequest.func37
github.com/authelia/authelia/v4/internal/middlewares/strip_path.go:22       handleRouter.StripPath.func38.1
github.com/valyala/fasthttp@v1.51.0/server.go:2359                          (*Server).serveConn
github.com/valyala/fasthttp@v1.51.0/workerpool.go:224                       (*workerPool).workerFunc
github.com/valyala/fasthttp@v1.51.0/workerpool.go:196                       (*workerPool).getCh.func1
runtime/asm_amd64.s:1650                                                    goexit"
```

**Reply by cyberb** [reply]
payload looks ok:

```
{"username":"test","password":"test1234","keepMeLoggedIn":false,"targetURL":"https://transmission.***.syncloud.it/"}
```

**Reply by james-d-elliott** [reply]
Unlikely to be related.. we can log the data to be sure. It's more likely to be the data (complete or otherwise) isn't making it to the socket for some unknown reason.  Are there any proxy logs?

**Reply by cyberb** [reply]
ok I found it as I tried to reuse nginx location example to also serve the portal (could not find a proper example on authrequest and portal under a subpath nginx example) this directive was there:
```
proxy_pass_request_body off;
```
probably I need separate locations for authrequest and portal.
